source/common/stats/heap_stat_data.cc

```diff
@@ -61,7 +73,145 @@ void HeapStatDataAllocator::debugPrint() {
 }
 #endif
 
-template class StatDataAllocatorImpl<HeapStatData>;
+class CounterImpl : public Counter, public MetricImpl {
+public:
+  CounterImpl(HeapStatData& data, HeapStatDataAllocator& alloc,
+              absl::string_view tag_extracted_name, const std::vector<Tag>& tags)
+      : MetricImpl(tag_extracted_name, tags, alloc.symbolTable()), data_(data), alloc_(alloc) {}
+
+  ~CounterImpl() override {
+    // MetricImpl must be explicitly cleared() before destruction, otherwise it
+    // will not be able to access the SymbolTable& to free the symbols. An RAII
+    // alternative would be to store the SymbolTable reference in the
+    // MetricImpl, costing 8 bytes per stat.
+    MetricImpl::clear();
+    alloc_.free(data_);
+  }
+
+  // Stats::Counter
+  void add(uint64_t amount) override {
+    data_.value_ += amount;
+    data_.pending_increment_ += amount;
+    data_.flags_ |= Flags::Used;
+  }
+  void inc() override { add(1); }
+  uint64_t latch() override { return data_.pending_increment_.exchange(0); }
+  void reset() override { data_.value_ = 0; }
+  bool used() const override { return data_.flags_ & Flags::Used; }
+  uint64_t value() const override { return data_.value_; }
+
+  SymbolTable& symbolTable() override { return alloc_.symbolTable(); }
+  StatName statName() const override { return data_.statName(); }
+
+private:
+  HeapStatData& data_;
+  HeapStatDataAllocator& alloc_;
```

Comment:
thought for consideration: If we are going all nutso on saving memory, 8 bytes for a reference to an effectively singleton allocator seems not worth it to me. I wonder if we should consider some type of scoped singleton here just for this case, same for gauge below. If you like this idea maybe a TODO/comment for later consideration?

Reply:
I think this 8 bytes is recoverable with a lot of work, but alas it is not a singleton. IsolatedStatsStore makes its own HeapStatDataAllocator object, so there can be lots of allocators in a process.

The recovery of this 8 bytes involves passing the allocator to lots of class methods that currently don't require them, notably having a private destructor and an explicit free(), but also Counter::name(), which is probably used a lot in tests. It's possible but quite a bit of work.  There's a bunch of other things to do which are easier, with more potential benefit.

Reply:
Sounds good